starky/src/f5g.rs

```diff
@@ -0,0 +1,997 @@
+#![allow(dead_code)]
+use core::ops::{Add, AddAssign, Div, DivAssign, Mul, MulAssign, Neg, Sub, SubAssign};
+use plonky::field_gl::Fr;
+use plonky::Field;
+use std::fmt::write;
+use std::hash::{Hash, Hasher};
+use std::slice;
+
+use core::fmt::{Display, Formatter};
+/// Irreducible polynomial: x5-3
+#[repr(C)]
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Default)]
+pub struct F5G {
+    pub cube: [Fr; 5],
+    pub dim: usize,
+}
+
+impl Hash for F5G {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.as_bytes().hash(state);
+        self.dim.hash(state);
+    }
+}
+
+impl F5G {
+    pub fn new(a: Fr, b: Fr, c: Fr, d: Fr, e: Fr) -> Self {
+        F5G {
+            cube: [a, b, c, d, e],
+            dim: 5,
+        }
+    }
+
+    #[inline(always)]
+    pub fn to_be(&self) -> Fr {
+        assert_eq!(self.dim, 1);
+        self.as_elements()[0]
+    }
+
+    #[inline(always)]
+    pub fn as_elements(&self) -> Vec<Fr> {
+        let elements = &[self.cube];
+        let ptr = elements.as_ptr();
+        let len = elements.len() * self.dim;
+        let elems: &[Fr] = unsafe { slice::from_raw_parts(ptr as *const Fr, len) };
+        elems.to_vec()
+    }
+
+    #[inline]
+    pub fn mul_scalar(self, b: usize) -> Self {
+        let b = Fr::from(b as u64);
+        let elems = self.as_elements();
+        if self.dim == 1 {
+            Self::from(elems[0] * b)
+        } else {
+            Self::new(
+                elems[0] * b,
+                elems[1] * b,
+                elems[2] * b,
+                elems[3] * b,
+                elems[4] * b,
+            )
+        }
+    }
+
+    #[inline]
+    fn eq(self, rhs: &Self) -> bool {
+        if self.dim == rhs.dim {
+            self.cube == rhs.cube
+        } else {
+            if self.dim == 1 {
+                self.cube[0] == rhs.cube[0]
+                    && rhs.cube[1] == Fr::ZERO
+                    && rhs.cube[2] == Fr::ZERO
+                    && rhs.cube[3] == Fr::ZERO
+                    && rhs.cube[4] == Fr::ZERO
+            } else {
+                self.cube[0] == rhs.cube[0]
+                    || (self.cube[1] == Fr::ZERO)
+                    || (self.cube[2] == Fr::ZERO)
+                    || (self.cube[3] == Fr::ZERO)
+                    || (self.cube[4] == Fr::ZERO)
+            }
+        }
+    }
+
+    #[inline]
+    pub fn gt(self, rhs: &Self) -> bool {
+        assert_eq!(self.dim, rhs.dim); // FIXME: align with JS
+        let les = self.as_elements();
+        let res = rhs.as_elements();
+        match self.dim {
+            5 => {
+                (les[0] > res[0]
```

Comment:
If les[0] > res[0], we don't have to care about the left limbs, same as the i-th limb comparsion.  BTW, there si a bug in f3g.  When comparing the second limb,  the expresion should be `                    || ((les[0] == res[0]) && (les[1] > res[1]))
`